rhp/v4/client.go

```diff
@@ -0,0 +1,666 @@
+	o := defaultOptions
+	for _, opt := range opts {
+		opt(&o)
+	}
+	c := &Client{
+		addr:        addr,
+		hostKey:     hostKey,
+		dialTimeout: o.DialTimeout,
+		rpcTimeout:  o.RPCTimeout,
+	}
+	return c, c.resetTransport(ctx)
+}
+
+// Close closes the client and waits for any spawned goroutines to finish.
+func (c *Client) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	var err error
+	if c.mux != nil {
+		err = c.mux.Close()
+		c.mux = nil
+	}
+	c.wg.Wait()
+	return err
+}
+
+// do blockingly performs an RPC. If retry is 'true', it will recursively retry
+// performing the rpc once if the error returned by rpcFn indicates a failure of
+// the underlying transport. This method is not intended to be called directly.
+// Instead 'withStream' should be called instead.
+func (c *Client) do(ctx context.Context, retry bool, rpcFn func(s *mux.Stream) error) error {
+	// dial a stream with a sane deadline
+	c.mu.Lock()
+	s := c.mux.DialStream()
+	c.mu.Unlock()
+	if err := s.SetDeadline(time.Now().Add(c.rpcTimeout)); err != nil {
+		return fmt.Errorf("failed to set deadline: %w", err)
+	}
+	defer s.Close()
+
+	// perform rpc
+	errRPC := rpcFn(s)
+
+	// check the error code to see if the error is part of the protocol or
+	// transport related - if it is transport related, reset the transport
+	// and retry the RPC
+	if retry && rhpv4.ErrorCode(errRPC) == rhpv4.ErrorCodeTransport {
+		c.mu.Lock()
+		if time.Since(c.lastMuxReset) < 10*time.Second {
+			// don't reset too often
+			c.mu.Unlock()
+			return fmt.Errorf("not resetting mux since it was recently reset: %w", errRPC)
+		} else if err := c.resetTransport(ctx); err != nil {
+			// failed to reset transport
+			c.mu.Unlock()
+			return fmt.Errorf("%v: %w", err, errRPC)
+		}
+		c.mu.Unlock()
+		return c.do(ctx, false, rpcFn) // retry
+	}
+	return errRPC
+}
+
+// withStream creates a stream for the caller to perform an RPC on. It will
+// write the RPC id and then leave it to the caller to perform the remaining RPC
+// in the callback. The stream will be closed automatically and any panics which
+// might occurr will be returned as an error.
+// NOTE: rpcFn might be called up to 2 times so it should be written in a way
+// that avoids carrying state over from the first call to the second.
+func (c *Client) withStream(ctx context.Context, rpcFn func(*mux.Stream) error) error {
+	done := make(chan struct{})
+	var doErr error
+	var s *mux.Stream
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		defer close(done)
+
+		// defer recover
+		defer func() {
+			if r := recover(); r != nil {
+				doErr = fmt.Errorf("a panic occurred while executing the rpc: %v", r)
+			}
+		}()
+
+		// perform RPC
+		doErr = c.do(ctx, true, rpcFn)
+	}()
+	select {
+	case <-ctx.Done():
+		// Caller interrupted the RPC - optimistically set deadline to abort
+		// goroutine as soon as possible
+		s.SetDeadline(time.Now())
```

Comment:
I think `s.Close` will be called anyway since it's deferred.
The reason why I called it optimistic is because I know that it won't affect ongoing connections but it will fail the next Read/Write on the stream. 
But I guess that's redundant considering that we close it anyway.